### PR TITLE
Gravity postprocessor

### PR DESCRIPTION
**Pull request by LudovicJnnt**

### CODE REVIEW

File: source/postprocess/satellite_gravity.cc [hunk 1]
```diff
@@ -0,0 +1,362 @@
...
+                      if (cell->is_locally_owned())
+                        {
+                          fe_values.reinit (cell);
+                          const std::vector<Point<dim> > &position_point = fe_values.get_quadrature_points();
+
+                          for (unsigned int q = 0; q < number_quadrature_points_cell; ++q)
+                            {
+                              const double dist=(position_satellite[0]-position_point[q][0])*(position_satellite[0]-position_point[q][0]) +  
+                                                (position_satellite[1]-position_point[q][1])*(position_satellite[1]-position_point[q][1]) +  
+                                                (position_satellite[2]-position_point[q][2])*(position_satellite[2]-position_point[q][2]);
```
Comment (gassmoeller) [inline source/postprocess/satellite_gravity.cc]: you can replace this by `(position_satellite - position_point).norm()`, easier to read :smile:

File: source/postprocess/satellite_gravity.cc [hunk 2]
```diff
@@ -0,0 +1,362 @@
...
+                          for (unsigned int q = 0; q < number_quadrature_points_cell; ++q)
+                            {
+                              const double dist=(position_satellite[0]-position_point[q][0])*(position_satellite[0]-position_point[q][0]) +  
+                                                (position_satellite[1]-position_point[q][1])*(position_satellite[1]-position_point[q][1]) +  
+                                                (position_satellite[2]-position_point[q][2])*(position_satellite[2]-position_point[q][2]);  
+
+                              double KK = G * density_all[number_cell_mpi * number_quadrature_points_cell + q] * fe_values.JxW(q) / pow(std::sqrt(dist),3);
+                              local_gx += KK * (position_satellite[0]-position_point[q][0]);  
+                              local_gy += KK * (position_satellite[1]-position_point[q][1]);  
+                              local_gz += KK * (position_satellite[2]-position_point[q][2]);
```
Comment (gassmoeller) [inline source/postprocess/satellite_gravity.cc]: you can write this as `g = KK * (position_satellite - position_point)`

File: source/postprocess/satellite_gravity.cc [hunk 3]
```diff
@@ -0,0 +1,362 @@
...
+    GravityCalculation<dim>::declare_parameters (ParameterHandler &prm)
+    {
+      prm.enter_subsection ("Postprocess");
+      {
+        prm.enter_subsection ("Gravity calculation");
+        {
+          prm.declare_entry ("Mapping spherical spacing", "1",
+                             Patterns::Double ());
+          prm.declare_entry ("Profile depth spacing", "1",
+                             Patterns::Double ());
```
Comment (gassmoeller) [inline source/postprocess/satellite_gravity.cc]: If you put a number in like this: `Double(0.0)` it means the input parameter has to be larger than 0.0

File: source/postprocess/satellite_gravity.cc [hunk 4]
```diff
@@ -0,0 +1,362 @@
...
+          prm.declare_entry ("Profile longitude", "0",
+                             Patterns::Double ());
+          prm.declare_entry ("Profile latitude", "0",
+                             Patterns::Double ());
+          prm.declare_entry ("Minimum longitude", "0",
+                             Patterns::Double ());
+          prm.declare_entry ("Minimum latitude", "0",
+                             Patterns::Double ());
+          prm.declare_entry ("Maximum longitude", "360",
```
Comment (gassmoeller) [inline source/postprocess/satellite_gravity.cc]: --> `Double(0.0,360.0)`

**Comment by tjhei** [comment]
Please remove all redundant files you accidentally added. Also fix the formatting using astyle.

**Comment by LudovicJnnt** [comment]
I have made all the changes Rene, Julianne and John asked. The postprocessor is by far quicker now.
the postprocessor allows the sphere, spherical shell and chunk model geometries. I let the quadrature degree +3 as it improves accuracy although just a little. 
Few points for discussions:
1-. quadrature degree - would it require its own declaration in .prm?
2.- inside the loop storing the density, position and JxW at quadrature points - I grouped density and JxW as I did not know how to store JxW separately (and it seems to work because JxW is only used with the density in this postprocessor)

**Comment by LudovicJnnt** [comment]
Please, this is ready for new review!

File: source/postprocess/gravity_point_values.cc [hunk 5]
```diff
@@ -0,0 +1,322 @@
...
+        << "# 7: gravity_x" << '\n'
+        << "# 8: gravity_y" << '\n'
+        << "# 9: gravity_z" << '\n'
+        << "# 10: gravity_norm" << '\n'
+        << "# 11: gravity_theory" << '\n'
+        << "# 12: gravity_diff (theoric-norm)" << '\n'
+        << "# 13: potential" << '\n'
+        << '\n';
+
+      // Temporary output during the build of this postprocessor WIP
```
Comment (gassmoeller) [inline source/postprocess/gravity_point_values.cc]: What means WIP? Did you include this only for debugging purposes? If so, please remove it. If it needs to stay in, it needs a better description of what you do here.

File: source/postprocess/gravity_point_values.cc [hunk 6]
```diff
@@ -0,0 +1,322 @@
...
+      const std::string filename2 = (this->get_output_directory() +
+                                     "checks_wip.txt");
+      std::ofstream f2 (filename2.c_str());
+      f2 << minimum_radius << ' ' << maximum_radius << '\n'
+         << minimum_longitude << ' ' << maximum_longitude << '\n'
+         << minimum_latitude << ' ' << maximum_latitude << '\n';
+
+      // Storing cartesian coordinate, density and JzW at quadrature points in a vector
```
Comment (gassmoeller) [inline source/postprocess/gravity_point_values.cc]: JzW -> JxW

File: source/postprocess/gravity_point_values.cc [hunk 7]
```diff
@@ -0,0 +1,322 @@
...
+      // avoids to use MaterialModel and fe_values within the loops.
+      unsigned int local_cell_number = (this->get_triangulation().n_locally_owned_active_cells());
```
Comment (gassmoeller) [inline source/postprocess/gravity_point_values.cc]: make this `const` and use a different variable as loop counter below, currently you use the same variable for different purposes. Also generally we name these variables `n_locally_owned_cells` not `locally_owned_cell_number`.

File: source/postprocess/gravity_point_values.cc [hunk 8]
```diff
@@ -0,0 +1,322 @@
...
+      const unsigned int number_quadrature_points_mpi = local_cell_number * number_quadrature_points_cell;
```
Comment (gassmoeller) [inline source/postprocess/gravity_point_values.cc]: this is not related to mpi, please remove that part. I would probably name this variable `n_local_quadrature_points`.

File: source/postprocess/gravity_point_values.cc [hunk 9]
```diff
@@ -0,0 +1,322 @@
...
+                  // gravity vector components x,y,z, and potential.
+                  cell = this->get_dof_handler().begin_active();
+                  local_cell_number = 0;
+                  for (; cell!=endc; ++cell)
+                    {
+                      if (cell->is_locally_owned())
+                        {
+                          for (unsigned int q = 0; q < number_quadrature_points_cell; ++q)
+                            {
+                              double dist = (position_satellite - position_point[local_cell_number * number_quadrature_points_cell + q]).norm();
```
Comment (gassmoeller) [inline source/postprocess/gravity_point_values.cc]: this can be `const double`

File: source/postprocess/gravity_point_values.cc [hunk 10]
```diff
@@ -0,0 +1,322 @@
...
+                              double KK = G * density_JxW[local_cell_number * number_quadrature_points_cell + q] / pow(dist,3);
```
Comment (gassmoeller) [inline source/postprocess/gravity_point_values.cc]: this as well `const double`

File: source/postprocess/gravity_point_values.cc [hunk 11]
```diff
@@ -0,0 +1,322 @@
...
+                              local_g += KK * (position_satellite - position_point[local_cell_number * number_quadrature_points_cell + q]);
+                              local_U -= G * density_JxW[local_cell_number * number_quadrature_points_cell + q] / dist;
+                            }
+                          ++local_cell_number;
+                        }
+                    }
+
+                  // assemble gravity component results from the mpi
```
Comment (gassmoeller) [inline source/postprocess/gravity_point_values.cc]: we use `assemble` only when assembling matrices. In this case I would describe it as `Sum gravity components over global domain`

File: source/postprocess/gravity_point_values.cc [hunk 12]
```diff
@@ -0,0 +1,322 @@
...
+          for (unsigned int i=0; i < number_points_longitude; ++i)
+            {
+              satellite_coordinate[1] = (minimum_longitude + ((maximum_longitude - minimum_longitude) / number_points_longitude) * i) * numbers::PI / 180.;
+
+              // loop on theta - satllite position [ , , theta]
+              for (unsigned int j=0; j < number_points_latitude; ++j)
+                {
+                  satellite_coordinate[2] = (minimum_latitude + ((maximum_latitude - minimum_latitude) / number_points_latitude) * j) * numbers::PI / 180.;
+                  Tensor<1,dim> local_g;
+                  double local_U = 0;
```
Comment (gassmoeller) [inline source/postprocess/gravity_point_values.cc]: What is U? Is it the gravity potential? then name it `local_gravity_potential`. There are too many letters in too many different scientific fields to remember them all (also there is often some ambiguity). I am not so much concerned about the `g` above because that is really universal.

File: source/postprocess/gravity_point_values.cc [hunk 13]
```diff
@@ -0,0 +1,322 @@
...
+                              local_U -= G * density_JxW[local_cell_number * number_quadrature_points_cell + q] / dist;
+                            }
+                          ++local_cell_number;
+                        }
+                    }
+
+                  // assemble gravity component results from the mpi
+                  const Tensor<1,dim> g
+                    = Utilities::MPI::sum (local_g, this->get_mpi_communicator());
+                  const double U
```
Comment (gassmoeller) [inline source/postprocess/gravity_point_values.cc]: global_gravity_potential?

File: source/postprocess/gravity_point_values.cc [hunk 14]
```diff
@@ -0,0 +1,322 @@
...
+                    = Utilities::MPI::sum (local_U, this->get_mpi_communicator());
+
+                  // analytical solution to calculate the theoritical gravity from a uniform desnity model.
```
Comment (gassmoeller) [inline source/postprocess/gravity_point_values.cc]: theoritical -> theoretical
desnity -> density

File: source/postprocess/gravity_point_values.cc [hunk 15]
```diff
@@ -0,0 +1,322 @@
...
+          prm.declare_entry ("Maximum radius", "0",
+                             Patterns::Double (0.0));
+          prm.declare_entry ("Minimum longitude", "0",
+                             Patterns::Double (0.0,360.0));
+          prm.declare_entry ("Minimum latitude", "0",
+                             Patterns::Double (0.0,180.0));
+          prm.declare_entry ("Maximum longitude", "360",
+                             Patterns::Double (0.0,360.0));
+          prm.declare_entry ("Maximum latitude", "180",
+                             Patterns::Double (0.0,180.0));
```
Comment (gassmoeller) [inline source/postprocess/gravity_point_values.cc]: You have removed the explanation from all of these parameters, but that means they will not be explained in the manual at all. Please put the descriptions back in, as is done in all other postprocessors (just add another argument to every call to `declare_entry` and copy the comment for the particular variable from the header file).

File: source/postprocess/gravity_point_values.cc [hunk 16]
```diff
@@ -0,0 +1,322 @@
...
+namespace aspect
+{
+  namespace Postprocess
+  {
+    ASPECT_REGISTER_POSTPROCESSOR(GravityPointValues,
+                                  "gravity calculation",
+                                  "A postprocessor that computes gravity and gravity potential "
+                                  "for a set of points (e.g. satellites) in or above the model "
+                                  "surface for a user-defined range of latitudes, longitudes "
+                                  "and radius. ")
```
Comment (gassmoeller) [inline source/postprocess/gravity_point_values.cc]: Remove the space after the documentation.

File: tests/gravity_point_values.prm [hunk 17]
```diff
@@ -0,0 +1,69 @@
+# A simple setup for gravity acceleration calculated at a point P
+# manual for more information.
```
Comment (gassmoeller) [inline tests/gravity_point_values.prm]: at a point P manual for more information 
-> 
at a point P. See the corresponding section in the manual for more information.

**Comment by gassmoeller** [comment]
You will need to update the test output. Apart from that this is ready.

**Comment by LudovicJnnt** [comment]
Test output updated. Now ready.
